### PR TITLE
Implement `.by`/`by`

### LINKED ISSUE #399: Implement `.by`
See https://github.com/tidyverse/dplyr/pull/6528

### PR DESCRIPTION
**Pull request by markfairbanks**
Closes #399 

**Overview of changes**
* Internal helper `compute_by()` was implemented (modeled after `dplyr`/`dbplyr` versions)
* A `by` arg was added to the `step_subset_i()`/`step_subset_j()` internal helpers.
  These are passed the result of `compute_by()` by `filter()`/`mutate()`/`summarize()`/`slice_*()`.

### CODE REVIEW

File: R/by.R [hunk 1]
```diff
@@ -0,0 +1,80 @@
+#' Helper for consistent documentation of `.by`
```
Comment (hadley) [inline R/by.R:1]: Do you need this here or can you import from what dplyr provides?

File: DESCRIPTION [hunk 2]
```diff
@@ -18,6 +18,7 @@ BugReports: https://github.com/tidyverse/dtplyr/issues
 Depends: 
     R (>= 3.3)
 Imports:
+    cli (>= 3.4.0),
     crayon,
     data.table (>= 1.13.0),
     dplyr (>= 1.0.3),
```
Comment (hadley) [inline DESCRIPTION:24]: Bump dplyr requirement?

File: R/by.R [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    error_call = error_call
+  )
+  names(out)
+}
+
+new_by <- function(uses_by, names) {
+  structure(list(uses_by = uses_by, names = names), class = "dtplyr_by")
+}
+
+default_by <- new_by(uses_by = FALSE, names = character())
```
Comment (hadley) [inline R/by.R:80]:
```suggestion
new_by <- function(uses_by = FALSE, names = character()) {
  structure(list(uses_by = uses_by, names = names), class = "dtplyr_by")
}
```

I think I find that bit easier to understand at a glance

File: R/step-mutate.R [hunk 4]
```diff
@@ -1,10 +1,14 @@
-step_mutate <- function(parent, new_vars = list(), use_braces = FALSE) {
+step_mutate <- function(parent, new_vars = list(), use_braces = FALSE, by = default_by) {
```
Comment (hadley) [inline R/step-mutate.R:1]:
```suggestion
step_mutate <- function(parent, new_vars = list(), use_braces = FALSE, by = new_by()) {
```

etc

File: R/step-subset-slice.R [hunk 5]
```diff
@@ -55,8 +55,9 @@
 #' # physical weight of the cars, so heavy cars are more likely to get
 #' # selected
 #' dt %>% slice_sample(weight_by = wt, n = 5)
-slice.dtplyr_step <- function(.data, ...) {
+slice.dtplyr_step <- function(.data, ..., .by = NULL) {
   dots <- capture_dots(.data, ..., .j = FALSE)
```
Comment (hadley) [inline R/step-subset-slice.R:59]: Might want to call `dplyr:::check_dot_by_typo(...)` here too.

File: R/step-subset-slice.R [hunk 6]
```diff
@@ -73,35 +74,38 @@ slice.dtplyr_step <- function(.data, ...) {
...
-  step_subset_i(.data, i = i)
+  step_subset_i(.data, i = i, by)
 }
 
 #' @rdname slice.dtplyr_step
 #' @importFrom dplyr slice_min
 #' @inheritParams dplyr::slice
 #' @export
-slice_min.dtplyr_step <- function(.data, order_by, ..., n, prop, with_ties = TRUE) {
+slice_min.dtplyr_step <- function(.data, order_by, ...,
```
Comment (hadley) [inline R/step-subset-slice.R:107]: If the arguments no longer fit on one line, our style to put each argument on its own line. https://github.com/lionel-/codegrip makes it easy to do so.

**Review by hadley — COMMENTED** [review]
Thanks for working on this!

**Comment by eutwt** [comment]
Sorry, won't have time to review this. Things are pretty hectic right now where I'm working

File: R/by.R (hunk 3 above)
Comment (markfairbanks) [reply R/by.R:66]: Ah yep that's much better

File: R/step-subset-slice.R (hunk 5 above)
Comment (markfairbanks) [reply R/step-subset-slice.R:47]: Looks like this is called in the generic in `dplyr` so the error works in `dtplyr`

**Comment by markfairbanks** [comment]
No worries, thanks @eutwt

File: R/step-mutate.R [hunk 7]
```diff
@@ -53,29 +63,7 @@ mutate_with_braces <- function(mutate_vars) {
...
-#'   * `"used"` retains only the columns used in `...` to create new
-#'     columns. This is useful for checking your work, as it displays inputs
-#'     and outputs side-by-side.
-#'   * `"unused"` retains only the columns _not_ used in `...` to create new
-#'     columns. This is useful if you generate new columns, but no longer need
-#'     the columns used to generate them.
-#'   * `"none"` doesn't retain any extra columns from `.data`. Only the grouping
-#'     variables and columns created by `...` are kept.
-#'
-#'  Note: With dtplyr `.keep` will only work with column names passed as symbols, and won't
```
Comment (hadley) [inline R/step-mutate.R:73]: Is this no longer true?

File: R/step-subset-slice.R (hunk 5 above)
Comment (hadley) [reply R/step-subset-slice.R:47]: Oh yeah, of course.

**Review by hadley — APPROVED** [review]
Thanks! If you want to merge this, I'll start the release process tomorrow morning.

File: R/step-mutate.R (hunk 7 above)
Comment (markfairbanks) [reply R/step-mutate.R:73]: Ah whoops yep it is. I was wondering why we had that separate from dplyr's docs.

**Pull request merged**